crates/astria-gossipnet/src/network_stream.rs

```diff
@@ -0,0 +1,271 @@
+use std::{
+    pin::Pin,
+    str::FromStr,
+    task::{
+        Context,
+        Poll,
+    },
+};
+
+use futures::StreamExt;
+#[cfg(feature = "dht")]
+use libp2p::kad::{
+    record::Key,
+    {
+        KademliaEvent,
+        QueryResult,
+    },
+};
+#[cfg(feature = "mdns")]
+use libp2p::mdns;
+use libp2p::{
+    gossipsub::{
+        self,
+        Message,
+        TopicHash,
+    },
+    kad::{
+        AddProviderError,
+        Addresses,
+        GetClosestPeersError,
+        GetProvidersError,
+    },
+    swarm::SwarmEvent,
+    Multiaddr,
+    PeerId,
+};
+use tracing::{
+    debug,
+    warn,
+};
+
+use crate::network::{
+    GossipnetBehaviourEvent,
+    Network,
+};
+
+#[derive(Debug)]
+pub enum Event {
+    NewListenAddr(Multiaddr),
+    Message(Message),
+    #[cfg(feature = "mdns")]
+    MdnsPeersConnected(Vec<PeerId>),
+    #[cfg(feature = "mdns")]
+    MdnsPeersDisconnected(Vec<PeerId>),
+    PeerConnected(PeerId),
+    PeerSubscribed(PeerId, TopicHash),
+    #[cfg(feature = "dht")]
+    FoundProviders(Option<Key>, Option<Vec<PeerId>>),
+    #[cfg(feature = "dht")]
+    GetProvidersError(GetProvidersError),
+    #[cfg(feature = "dht")]
+    Providing(Key),
+    #[cfg(feature = "dht")]
+    ProvideError(AddProviderError),
+    #[cfg(feature = "dht")]
+    RoutingUpdated(PeerId, Addresses),
+    #[cfg(feature = "dht")]
+    FoundClosestPeers(Vec<PeerId>),
+    #[cfg(feature = "dht")]
+    GetClosestPeersError(GetClosestPeersError),
+}
+
+impl futures::Stream for Network {
+    type Item = Event;
+
+    fn poll_next(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
```

Comment:
I think this could be a stream returning `type Item = Result<Event, Error>;` and let the consumer of the network handle the errors (like logging).